certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,240 @@
+*/
+invariant lengthConsistency()
+    length() == absolute_diff(to_mathint(end()), to_mathint(begin()))
+    {
+        preserved {
+            requireInvariant boundariesConsistency;
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: out of bounds is always 0                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant outOfBoundsDefault(int128 key)
+    (
+        key < begin() &&  // Is before beginning
+        key >= end() // Is after ending
+    ) => at_(to_uint256(key)) == 0
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: empty() is length 0 and no element exists                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant emptiness(int128 key)
+    empty() <=> length() == 0 && at_(to_uint256(key)) == 0
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: front() and back() point to the [begin, end) range limits                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant pointerCorrectness()
+    back() == at_(to_uint256(begin())) && front() == at_(to_uint256(end()) - 1)
+    {
+        preserved {
+            requireInvariant boundariesConsistency;
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: pushFront adds an element at the beginning of the queue                                       │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushFront {
+    // Sanity
+    requireInvariant boundariesConsistency;
+    
+    bytes32 value;
+
+    int128 beginBefore = begin();
+    int128 beginNext = beginBefore - 1;
+    bytes32 frontBefore = front();
+
+    pushFront@withrevert(value);
+
+    int128 beginAfter = begin();
+    bytes32 frontAfter = front();
+
+    bool success = !lastReverted;
+
+    // liveness
+    assert success;
+
+    // effect
+    assert success => (
+        frontAfter == value && // Front set to value
+        beginAfter == beginNext // Begin is adjusted
+    ), "front set to value and begin is adjusted";
+
+    // no side effect
+    assert frontBefore == frontAfter <=> frontBefore == value, "front doesn't change only if it was already set";
```

Comment:
It would be nice to check that `end()` and `back()` are not affected.

Reply:
Regarding `end()`, I don't think it's necessary since it's not part of the interface as we spoke last Friday.

Also, I ended up with several rules that prove `end` and `back` are not affected.

- `lengthConsistency` guarantees length is always `end() - begin()`
- `noLengthChange`  guarantees length is only changed by 1
- `noDataChange` proves the only operations that can change the values in the queue
- `(push/pop)(Front/Back)Consistency` prove that no values in the queue were modified

Therefore, there's no way back is affected, and even if `end` is affected, is guaranteed to be consistent.